be/src/formats/parquet/file_writer.cpp

```diff
@@ -177,139 +192,718 @@ void ParquetBuildHelper::build_compression_type(::parquet::WriterProperties::Bui
+    auto inner_node = mid_node->field(0);
+
+    unsigned char* nulls = nullptr;
+    if (col->is_nullable()) {
+        const auto null_column = down_cast<NullableColumn*>(col.get())->null_column();
+        nulls = null_column->get_data().data();
+    }
+
+    const auto array_column = down_cast<ArrayColumn*>(ColumnHelper::get_data_column(col.get()));
+    const auto elements = array_column->elements_column();
+    const auto offsets = array_column->offsets_column()->get_data();
+    Context child_ctx(ctx._max_def_level + node->is_optional() + 1, ctx._max_rep_level + 1, ctx.size() + elements->size());
+
+    int subcol_size = 0;
+    for (auto i = 0; i < ctx.size(); i++) {
+        auto [idx, def_level, rep_level] = ctx.get(i);
+        if (idx == Context::kNULL || nulls[idx]) {
+            child_ctx.append(Context::kNULL, def_level, rep_level);
+            continue;
+        }
+        auto array_size = offsets[idx + 1] - offsets[idx];
+        if (array_size == 0) {
+            child_ctx.append(Context::kNULL, def_level + node->is_optional(), rep_level);
+            continue;
         }
+        child_ctx.append(subcol_size++, def_level + node->is_optional() + 1, rep_level);
+        for (auto offset = 1; offset < array_size; offset++) {
+            child_ctx.append(subcol_size++, def_level + node->is_optional() + 1, child_ctx._max_rep_level);
+        }
+    }
+
+    DCHECK(elements->size() == subcol_size);
+    return _add_column_chunk(child_ctx, type_desc.children[0], inner_node, elements);
+}
+
+Status FileWriterBase::_add_map_column_chunk(Context& ctx, const TypeDescriptor& type_desc,
+                                             const ::parquet::schema::NodePtr& node, const ColumnPtr& col) {
+    DCHECK(type_desc.type == TYPE_MAP);
+    auto outer_node = std::static_pointer_cast<::parquet::schema::GroupNode>(node);
+    auto mid_node = std::static_pointer_cast<::parquet::schema::GroupNode>(outer_node->field(0));
+    auto key_node = mid_node->field(0);
+    auto value_node = mid_node->field(1);
+
+    unsigned char* nulls = nullptr;
+    if (col->is_nullable()) {
+        const auto null_column = down_cast<NullableColumn*>(col.get())->null_column();
+        nulls = null_column->get_data().data();
+    }
 
-        const auto type = _output_expr_ctxs[i]->root()->type().type;
-        switch (type) {
-        case TYPE_BOOLEAN: {
-            DISPATCH_PARQUET_NUMERIC_WRITER(BoolWriter, starrocks::BooleanColumn, bool)
-            break;
+    const auto map_column = down_cast<MapColumn*>(ColumnHelper::get_data_column(col.get()));
+    const auto keys = map_column->keys_column();
+    const auto values = map_column->values_column();
+    const auto offsets = map_column->offsets_column()->get_data();
+
+    Context key_ctx(ctx._max_def_level + node->is_optional(), ctx._max_rep_level + 1, ctx.size() + keys->size());
+    Context value_ctx(ctx._max_def_level + node->is_optional(), ctx._max_rep_level + 1, ctx.size() + values->size());
```

Comment:
Seems no need to have two `Context` for key and value, since they are the same.

Reply:
done